### DISCUSSION TITLE
Expressions and constants

### DISCUSSION THREAD

**Discussion by ralsina** [dune3d/dune3d#83] [Ideas]
I am playing with dune3d and one thing I miss from freecad is the ability to have measurements defined outside the constraint itself, what freecad calls "spreadsheets". 

So, for example I can define "width" and use that in multiple places, then change it in one place and have the model reflect that.

I am an experienced python programmer so I am willing to take a shot at this if it has a chance of getting accepted

**Comment by carrotIndustries** [comment]
Yeah, I've though of that before, braindump incoming in 3... 2... 1!

The solvespace solver that's used to solve sketches operates on equations and parameters. Entities such as lines add parameters for their points and constraints add equations describing the relationships among the parameters. So adding support for user-defined equations and parameters is certainly within in the realm of possibility. Here's how it could work from a user's and implementation point of view:

For each group, users can add add their own parameters. Parameters have an immutable and unique name, a value and can be either fixed or be modifiable by the solver. Rather than just accepting numbers, constraints that have a value now take expressions. An expression could be something like `foo*2.5 + 1`. While the user enters these as strings and they're stored as such, they're also immediately converted to an abstract representation for easy use later on. Some possible representations:
 - An `Expr` class similar to the one from solvespace. We can't use the one from solvespace directly since the usage of solvespace code should be restricted to the `System` class. It also uses the special arena allocator that makes keeping them around rather difficult.
 - A stack machine

Some constraints are directly mapped to the ones from the solvespacer solver, while for others, we directly emit the required equations. Since the solvespace constraints only support numbers and not expressions, we now have to take care of the equations for these ones as well. In that step we also convert expressions from the above representation to solvespace's `Expr`.

Similar to freecad's spreadsheet, it probably makes sense being able to add equations on the group level, not tied to any constraint.

The biggest task I think is to define the grammar(?) for the expressions, how to refer to parameters, etc and implement the parser.

Happy hacking!

**Reply by ralsina** [reply]
If you don't mind it being insecure we can just use something like python's eval for the grammar.

In fact it's quite nice, since you can pass it a dictionary that it will use to resolve symbols.

**Reply by carrotIndustries** [reply]
> If you don't mind it being insecure we can just use something like python's eval for the grammar.

Unfortunately, I don't really want that. Also, python is a bit of a heavy dependency for this purpose.
Apart from that, the problem is that eval will just give us the value of the expression. We need the input parsed into something we can ultimately convert into solvespace's `Expr`.

**Comment by ruevs** [comment]
For cross reference here is the corresponding SolveSpace issue https://github.com/solvespace/solvespace/issues/77. It is one of the oldest and most requested. There is a lot of interesting discussion there including from @jwesthues -  the original author of SolveSpace.

Of note are the most recent comments discussing units vs. dimensionless numbers/parameters. It is something worth thinking about early on.

Also interesting are the two proof of concept implementations and videos demonstrating them.

Here's is one of the better thought out more recent duplicates of this popular request https://github.com/solvespace/solvespace/issues/542. The first point in our [pinned issue](https://github.com/solvespace/solvespace/issues/1024) has links to all similar requests over the years.

**Reply by ralsina** [reply]
I have a long weekend coming, no promises but I can take a shot

**Comment by drwho495** [comment]
any updates on this?

**Comment by JaneSmith** [comment]
Is there any way to speed up the implementation of this? e.g. donation? I'm so missing this feature. Dune3D is such a joy to work with compared to SolveSpace and FreeCAD, with the one exception of this issue.

**Comment by xerio123** [comment]
i will make it

**Comment by seedship** [comment]
Any updates on this? :eyes: +1 to what @JaneSmith said
